File: src/backend/go/internal/service/file_service.go
package service

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"io"
	"math"
	"os"
	"path/filepath"
	"strings"

	"github.com/todo/backend/go/internal/config"
	"github.com/todo/backend/go/internal/dto"
	"github.com/todo/backend/go/internal/models"
	"github.com/todo/backend/go/internal/repository"
	"gorm.io/gorm"
)

// ErrFileNotFound is returned when a requested file does not exist.
var ErrFileNotFound = errors.New("file not found")

// ErrFileTooLarge is returned when an uploaded file exceeds the configured maximum size.
var ErrFileTooLarge = errors.New("file exceeds the maximum allowed upload size")

// ErrFileContentMissing is returned when a file's metadata exists but its content is missing on disk.
var ErrFileContentMissing = errors.New("file content not found on disk")

// UploadInput carries the raw data needed to persist an uploaded file - decouples
// the service from Gin's multipart.FileHeader type.
type UploadInput struct {
	OriginalName string
	ContentType  string
	Size         int64
	Reader       io.Reader
}

// DownloadTarget carries everything needed to stream a file back to the client.
type DownloadTarget struct {
	Path        string
	Name        string
	ContentType string
}

// FileService defines the business-logic contract for uploaded files.
// Mirrors IFileService in C#.
type FileService interface {
	GetAll(page, pageSize int) (dto.PaginatedResponse[dto.FileResponse], error)
	GetByID(id uint) (dto.FileResponse, error)
	Upload(input UploadInput) (dto.FileResponse, error)
	GetDownloadTarget(id uint) (DownloadTarget, error)
	Delete(id uint) error
}

type fileService struct {
	repo          repository.FileRepository
	storageDir    string
	maxUploadSize int64
}

// NewFileService constructs the service with its repository dependency injected.
func NewFileService(repo repository.FileRepository, cfg *config.Config) FileService {
	return &fileService{
		repo:          repo,
		storageDir:    cfg.FileStoragePath,
		maxUploadSize: cfg.MaxUploadSizeBytes,
	}
}

// ── Mapping ───────────────────────────────────────────────────────────────────

func fileToResponse(m *models.File) dto.FileResponse {
	return dto.FileResponse{
		ID:          m.ID,
		Name:        m.Name,
		Extension:   m.Extension,
		Size:        m.Size,
		ContentType: m.ContentType,
		CreatedAt:   m.CreatedAt,
		UpdatedAt:   m.UpdatedAt,
	}
}

func toFilePaginated(result repository.FilePaginatedResult, page, pageSize int) dto.PaginatedResponse[dto.FileResponse] {
	items := make([]dto.FileResponse, len(result.Items))
	for i := range result.Items {
		items[i] = fileToResponse(&result.Items[i])
	}
	return dto.PaginatedResponse[dto.FileResponse]{
		Items:      items,
		Total:      result.Total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: int(math.Ceil(float64(result.Total) / float64(pageSize))),
	}
}

func (s *fileService) getOrNotFound(id uint) (*models.File, error) {
	file, err := s.repo.FindByID(id)
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, ErrFileNotFound
	}
	return file, err
}

// ── Queries ───────────────────────────────────────────────────────────────────

func (s *fileService) GetAll(page, pageSize int) (dto.PaginatedResponse[dto.FileResponse], error) {
	result, err := s.repo.FindAll((page-1)*pageSize, pageSize)
	if err != nil {
		return dto.PaginatedResponse[dto.FileResponse]{}, err
	}
	return toFilePaginated(result, page, pageSize), nil
}

func (s *fileService) GetByID(id uint) (dto.FileResponse, error) {
	file, err := s.getOrNotFound(id)
	if err != nil {
		return dto.FileResponse{}, err
	}
	return fileToResponse(file), nil
}

// ── Commands ──────────────────────────────────────────────────────────────────

func (s *fileService) Upload(input UploadInput) (dto.FileResponse, error) {
	if input.Size > s.maxUploadSize {
		return dto.FileResponse{}, ErrFileTooLarge
	}

	// Strip any directory components from the client-supplied name to prevent path traversal.
	originalName := filepath.Base(input.OriginalName)
	if originalName == "" || originalName == "." || originalName == string(filepath.Separator) {
		originalName = "unnamed"
	}
	extension := strings.TrimPrefix(filepath.Ext(originalName), ".")

	if err := os.MkdirAll(s.storageDir, 0o755); err != nil {
		return dto.FileResponse{}, err
	}

	prefix, err := randomHex(16)
	if err != nil {
		return dto.FileResponse{}, err
	}
	// A random prefix avoids collisions/overwrites between uploads that share a name.
	location := filepath.Join(s.storageDir, prefix+"_"+originalName)

	out, err := os.Create(location)
	if err != nil {
		return dto.FileResponse{}, err
	}
	defer out.Close()

	written, err := io.Copy(out, input.Reader)
	if err != nil {
		return dto.FileResponse{}, err
	}

	var contentType *string
	if input.ContentType != "" {
		contentType = &input.ContentType
	}

	file := &models.File{
		Name:        originalName,
		Extension:   extension,
		Size:        written,
		ContentType: contentType,
		Location:    location,
	}
	created, err := s.repo.Create(file)
	if err != nil {
		return dto.FileResponse{}, err
	}
	return fileToResponse(created), nil
}

func (s *fileService) GetDownloadTarget(id uint) (DownloadTarget, error) {
	file, err := s.getOrNotFound(id)
	if err != nil {
		return DownloadTarget{}, err
	}
	if _, err := os.Stat(file.Location); err != nil {
		return DownloadTarget{}, ErrFileContentMissing
	}

	contentType := "application/octet-stream"
	if file.ContentType != nil && *file.ContentType != "" {
		contentType = *file.ContentType
	}
	return DownloadTarget{Path: file.Location, Name: file.Name, ContentType: contentType}, nil
}

func (s *fileService) Delete(id uint) error {
	file, err := s.getOrNotFound(id)
	if err != nil {
		return err
	}
	if err := s.repo.Delete(file); err != nil {
		return err
	}
	if _, statErr := os.Stat(file.Location); statErr == nil {
		return os.Remove(file.Location)
	}
	return nil
}

// ── Helpers ───────────────────────────────────────────────────────────────────

func randomHex(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

File: src/backend/dotnet/TodoApi/Services/FileService.cs
using Microsoft.AspNetCore.Http;
using TodoApi.Data;
using TodoApi.DTOs;
using TodoApi.Repositories;
using TodoShared.Models;

namespace TodoApi.Services;

public class FileService(IFileRepository repository, AppDbContext db, IConfiguration configuration) : IFileService
{
    private readonly string _storageDir = configuration["FileStorage:Path"] ?? "uploads";
    private readonly long _maxUploadSizeBytes = configuration.GetValue<long?>("FileStorage:MaxUploadSizeBytes") ?? 10 * 1024 * 1024;

    // ── Mapping ───────────────────────────────────────────────────────────────

    private static FileResponse ToResponse(FileEntity file) =>
        new(file.Id, file.Name, file.Extension, file.Size, file.ContentType, file.CreatedAt, file.UpdatedAt);

    private static PaginatedResponse<FileResponse> ToPaginated(
        IEnumerable<FileEntity> items, int total, int page, int pageSize) =>
        new(
            items.Select(ToResponse),
            total,
            page,
            pageSize,
            (int)Math.Ceiling(total / (double)pageSize)
        );

    private async Task<FileEntity> GetOrThrowAsync(int id, CancellationToken ct)
    {
        var file = await repository.GetByIdAsync(id, ct)
            ?? throw new KeyNotFoundException($"File {id} not found.");
        return file;
    }

    // ── Queries ───────────────────────────────────────────────────────────────

    public async Task<PaginatedResponse<FileResponse>> GetAllAsync(int page, int pageSize, CancellationToken ct = default)
    {
        var (items, total) = await repository.GetAllAsync((page - 1) * pageSize, pageSize, ct);
        return ToPaginated(items, total, page, pageSize);
    }

    public async Task<FileResponse> GetByIdAsync(int id, CancellationToken ct = default)
    {
        var file = await GetOrThrowAsync(id, ct);
        return ToResponse(file);
    }

    public async Task<FileDownloadTarget> GetDownloadTargetAsync(int id, CancellationToken ct = default)
    {
        var file = await GetOrThrowAsync(id, ct);
        if (!System.IO.File.Exists(file.Location))
        {
            throw new KeyNotFoundException($"File {id} content not found on disk.");
        }
        return new FileDownloadTarget(file.Location, file.Name, file.ContentType ?? "application/octet-stream");
    }

    // ── Commands ──────────────────────────────────────────────────────────────

    public async Task<FileResponse> UploadAsync(IFormFile file, CancellationToken ct = default)
    {
        if (file.Length > _maxUploadSizeBytes)
        {
            throw new FileTooLargeException($"File exceeds the maximum allowed size of {_maxUploadSizeBytes} bytes.");
        }

        // Strip any directory components from the client-supplied name to prevent path traversal.
        var originalName = Path.GetFileName(file.FileName);
        if (string.IsNullOrEmpty(originalName))
        {
            originalName = "unnamed";
        }
        var extension = Path.GetExtension(originalName).TrimStart('.').ToLowerInvariant();

        Directory.CreateDirectory(_storageDir);

        // A random prefix avoids collisions/overwrites between uploads that share a name.
        var storedName = $"{Guid.NewGuid():N}_{originalName}";
        var location = Path.GetFullPath(Path.Combine(_storageDir, storedName));

        await using (var stream = System.IO.File.Create(location))
        {
            await file.CopyToAsync(stream, ct);
        }

        var entity = new FileEntity
        {
            Name = originalName,
            Extension = extension,
            Size = file.Length,
            ContentType = file.ContentType,
            Location = location,
            CreatedAt = DateTime.UtcNow,
        };
        var created = await repository.AddAsync(entity, ct);
        await db.SaveChangesAsync(ct);
        return ToResponse(created);
    }

    public async Task DeleteAsync(int id, CancellationToken ct = default)
    {
        var file = await GetOrThrowAsync(id, ct);
        repository.Delete(file);
        await db.SaveChangesAsync(ct);
        try
        {
            if (System.IO.File.Exists(file.Location))
            {
                System.IO.File.Delete(file.Location);
            }
        }
        catch (IOException)
        {
            // Content already missing or inaccessible on disk - nothing left to clean up.
        }
    }
}
